Add connection limit to WebSocket server

shekharnwagh opened the pull request:
# Add connection limit to WebSocket server

## Summary

Adds configurable connection limits to the WebSocket server to prevent overload and maintain stability. When the limit is reached, new connections are rejected with a clear error message.

## Changes

- **Connection limit enforcement**: Rejects connections when `MAX_CONNECTIONS` is reached
- **Configurable via environment**: Set `VIP_RTC_WS_MAX_CONNECTIONS` (default: 1500, or -1 for unlimited)
- **Structured error responses**: Returns HTTP 503 with JSON body containing error code and message
- **Metrics tracking**: New Prometheus counter `wpvip_rtc_connection_limit_reached_total`

## Testing

Test with websocat and curl to verify:
- Connections are properly rejected when limit is reached
- Error response includes correct status code (503) and JSON body
- Error code and message are user-friendly and actionable

### Websocat

```sh
websocat "ws://localhost:1234/site-1/postType/post-{postId}?auth={JWT_TOKEN}"
```

### Curl

Useful to inspect the HTTP response body message.

```sh
curl -s -i \
	-H "Connection: Upgrade" \
	-H "Upgrade: websocket" \
	-H "Sec-WebSocket-Version: 13" \
	-H "Sec-WebSocket-Key: {WEBSOCKET_KEY}" \
	"http://localhost:1234/site-1/postType/post-{postId}?auth={JWT_TOKEN}"
```

shekharnwagh committed:
Add connection limit to WebSocket server

Rejects new connections when the limit is reached for server stability.
Connections are rejected during HTTP upgrade with a 503 status and
structured JSON error response.

The limit is configurable via VIP_RTC_WS_MAX_CONNECTIONS environment
variable and defaults to 1500 connections. Set to -1 for unlimited.

chriszarate commented on websocket-server/metrics.ts:
```diff
@@ -61,6 +61,11 @@ const connectionCloseCounter = new Counter( {
+const connectionLimitReachedCounter = new Counter( {
+	name: `${ METRICS_NAMESPACE }_connection_limit_reached_total`,
+	help: 'Total number of connections rejected due to connection limit',
+} );
```
I think a generic error counter might be better, with labels for different error types. Possibly can be combined with the auth failure counter.

chriszarate commented on websocket-server/index.ts:
```diff
@@ -116,6 +118,26 @@ server.on( 'upgrade', ( request: IncomingMessage, socket: Duplex, head: Buffer )
+				'Content-Type: application/json\r\n' +
+				'\r\n' +
+				errorResponse
+		);
```
HTTP1 text fun! :) 

Can this distinct error be detected on the client? So that we can display a message specific to the connection limit?

shekharnwagh replied:
As it happens the client can't read the response body, so changed to instead allowing the connection to be upgraded and then immediately closing with a custom WebSocket close code which is visible to the client.

Updated in 50af680

shekharnwagh committed:
Use WebSocket close code for connection limit rejection

Changes connection limit rejection from HTTP 503 response to WebSocket
close code 4002. This allows clients to receive the exact error reason
through the standard WebSocket close mechanism. This is required as
the exact HTTP response body is not available to the client.

This also consolidates all connection failure tracking into a single metric
counter with labeled reasons (auth failures and connection limit

chriszarate reviewed (approved):
Very nice!

shekharnwagh committed:
Merge branch 'trunk' into add/websocket-server-connection-limit
